Device manager client

cwillisf opened the pull request:
### Resolves

Progress toward #518 

### Proposed Changes

This change adds a new type of I/O device, called `DeviceManager`, which acts as a client for the Scratch Device Manager application. An instance of this new type is now available in the runtime's list of I/O devices and is called `deviceManager`.

### Test Coverage

A few basic tests for the new `DeviceManager` I/O device have been added as part of this change.

Testing the actual functionality requires a running instance of the Scratch Device Manager application. You can test this in the browser console with commands like this:

Retrieve a list of currently-available WeDo 2.0 devices (and trigger a scan if there isn't one already going):
```js
Scratch.vm.runtime.ioDevices.deviceManager.list('wedo2', 'test').then(x => console.log(`List result: ${x}`))
```

Create a `DeviceFinder` to handle searching for and connecting to a WeDo 2.0:
```js
var finder = Scratch.vm.runtime.ioDevices.deviceManager.searchAndConnect('test', 'wedo2');
finder.promise.then(x => console.log(`Result: ${x}`), reason => console.log(`Failed: ${reason}`));
```
You should start seeing list requests once per second in your browser's network panel.

If you connect a WeDo 2.0 through the Device Manager, you should see that the result is a socket. If you instead cancel the search:
```js
finder.cancel();
```
...you should see `Failed: canceled` in your console.

Commit:
Add an I/O device to represent the Device Manager

Commit:
Implement a few tests for io/deviceManager

Commit:
Remove test for canceling a search

When no Device Manager instance is present this test takes a long time
and can cause timeouts. We'll need to fake the HTTP requests if we want
to do this sort of test reliably.

Commit:
DeviceManager: add more jsdoc, remove `_sockets`

It turns out we don't need the Device Manager to centrally track each
socket, so this change removes the Device Manager's `_sockets` property
and related code.

cwillisf commented on src/io/deviceManager.js:
```diff
@@ -0,0 +1,335 @@
+
+        this._socket.on('deviceWasOpened', () => this.onDeviceWasOpened());
+        this._socket.on('disconnect', () => this.onDisconnect());
+        this._connectionTimeout = setTimeout(() => this.onTimeout(), DeviceOpener.CONNECTION_TIMEOUT_MS);
```
The `() => this.foo()` here is being used as a substitute for `foo.bind(this)`. I could use `bind` if you'd prefer -- I don't feel strongly either way.

rschamp replied:
Fine with me this way.  Using `bind` at run time rather than construct time is slightly less efficient, but not for something like this that won't be called frequently. I think the other thing you could do is use `bindAll` in the constructor for instance methods like we do in the GUI (where it's important to not `bind` in `render`). Then `this.onTimeout` should already be bound.

rschamp commented on package.json:
```diff
@@ -45,6 +46,7 @@
     "scratch-render": "latest",
     "scratch-storage": "latest",
     "script-loader": "0.7.0",
+    "socket.io-client": "^1.7.3",
```
Both of these should probably be explicitly pinned since they're used within the application.

rschamp commented on src/io/deviceManager.js:
```diff
@@ -0,0 +1,335 @@
+     * React to the socket becoming disconnected.
+     */
+    onDisconnect () {
+        this.removeSocket();
```
I don't see this defined anywhere, maybe left behind from a refactor?

Commit:
Respond to review comments

- Pin `got` and `socket-io.client`
- Finish removing `removeSocket`-related code

rschamp reviewed (changes requested):
@cwillisf I'm having trouble building this for some reason. I get warnings, which you can also see in the log for the Travis build:
```
    WARNING in ./~/ws/lib/Validation.js
    Module not found: Error: Can't resolve 'utf-8-validate' in '/Users/rschamp/Documents/Scratch/scratch-vm/node_modules/ws/lib'
     @ ./~/ws/lib/Validation.js 10:19-44
     @ ./~/ws/lib/Receiver.js
     @ ./~/ws/index.js
     @ ./~/engine.io-client/lib/transports/websocket.js
     @ ./~/engine.io-client/lib/transports/index.js
     @ ./~/engine.io-client/lib/socket.js
     @ ./~/engine.io-client/lib/index.js
     @ ./~/engine.io-client/index.js
     @ ./~/socket.io-client/lib/manager.js
     @ ./~/socket.io-client/lib/index.js
     @ ./src/io/deviceManager.js
     @ ./src/engine/runtime.js
     @ ./src/virtual-machine.js
     @ ./src/index.js

    WARNING in ./~/ws/lib/BufferUtil.js
    Module not found: Error: Can't resolve 'bufferutil' in '/Users/rschamp/Documents/Scratch/scratch-vm/node_modules/ws/lib'
     @ ./~/ws/lib/BufferUtil.js 10:19-40
     @ ./~/ws/lib/Sender.js
     @ ./~/ws/index.js
     @ ./~/engine.io-client/lib/transports/websocket.js
     @ ./~/engine.io-client/lib/transports/index.js
     @ ./~/engine.io-client/lib/socket.js
     @ ./~/engine.io-client/lib/index.js
     @ ./~/engine.io-client/index.js
     @ ./~/socket.io-client/lib/manager.js
     @ ./~/socket.io-client/lib/index.js
     @ ./src/io/deviceManager.js
     @ ./src/engine/runtime.js
     @ ./src/virtual-machine.js
     @ ./src/index.js
```

And then while that's happening, the GUI won't build with the vm linked:

```
ERROR in ../scratch-vm/dist/node/scratch-vm.js
Module not found: Error: Can't resolve 'fs' in '/Users/rschamp/Documents/Scratch/scratch-vm/dist/node'
 @ ../scratch-vm/dist/node/scratch-vm.js 1038:17-30
 @ ./src/containers/gui.jsx
 @ ./src/index.jsx
 @ multi (webpack)-dev-server/client?http://0.0.0.0:8601 ./src/index.jsx

ERROR in ../scratch-vm/dist/node/scratch-vm.js
Module not found: Error: Can't resolve 'net' in '/Users/rschamp/Documents/Scratch/scratch-vm/dist/node'
 @ ../scratch-vm/dist/node/scratch-vm.js 7283:17-31
 @ ./src/containers/gui.jsx
 @ ./src/index.jsx
 @ multi (webpack)-dev-server/client?http://0.0.0.0:8601 ./src/index.jsx

ERROR in ../scratch-vm/dist/node/scratch-vm.js
Module not found: Error: Can't resolve 'child_process' in '/Users/rschamp/Documents/Scratch/scratch-vm/dist/node'
 @ ../scratch-vm/dist/node/scratch-vm.js 34353:17-41
 @ ./src/containers/gui.jsx
 @ ./src/index.jsx
 @ multi (webpack)-dev-server/client?http://0.0.0.0:8601 ./src/index.jsx

ERROR in ../scratch-vm/dist/node/scratch-vm.js
Module not found: Error: Can't resolve 'tls' in '/Users/rschamp/Documents/Scratch/scratch-vm/dist/node'
 @ ../scratch-vm/dist/node/scratch-vm.js 34365:17-31
 @ ./src/containers/gui.jsx
 @ ./src/index.jsx
 @ multi (webpack)-dev-server/client?http://0.0.0.0:8601 ./src/index.jsx
Child html-webpack-plugin for "index.html":
    chunk    {0} index.html 541 kB [entry] [rendered]
        [0] ./~/lodash/lodash.js 540 kB {0} [built]
        [1] ./~/html-webpack-plugin/lib/loader.js!./src/index.ejs 627 bytes {0} [built]
        [2] (webpack)/buildin/global.js 509 bytes {0} [built]
        [3] (webpack)/buildin/module.js 517 bytes {0} [built]
webpack: Failed to compile.
```

Any ideas?  I found this, looking for the bufferutils warning: https://github.com/socketio/socket.io-client/issues/933

cwillisf commented:
The warnings are indeed related to this, I believe: https://github.com/socketio/socket.io-client/issues/933

However, I hadn't seen those GUI errors. I'll investigate.

Commit:
Fix socket.io-client dependency problems re: build

The Socket.io client library pulls in some extra dependencies in a way
that isn't fully compatible with Webpack. Switching to their build
output file, instead of the file specified in their `package.json`,
makes those errors go away.

cwillisf commented:
Using `require('socket.io-client/dist/socket.io')` fixes the build warnings in `scratch-vm` as well as the errors in `scratch-gui`.

@rschamp: you mentioned in a separate discussion that this change has increased the size of the `scratch-vm` build output quite a bit. That's true in the absolute sense (622 KB -> 973 KB) it's dwarfed by the rest of `scratch-gui`: `scratch-gui`'s `gui.js` build output increases by very roughly the same raw amount but a much lower percentage (7.2 MB -> 7.4 MB). I still think it's worth trying to decrease the build size but maybe not as a right-this-moment priority.
